---
title: "Mesangial RAGE KO Transcriptomic Analysis"
author: "Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    fig_width: 7
    fig_height: 7
theme: cosmo
---

Source: https://github.com/markziemann/mesangial_rageko

# Introduction

Looking at doing comparison of control and RAGE KO Mouse mesangial cells RNA-seq and miRNA-seq.

# Methods

* RAGE KO data sequenced at Baker. 61 nt length.

* WT data sequenced at BGI with 91 nt length.

* Quality trimming with skewer 0.2.2 to ensure 3' end base quality > 20.

* GENECODE mouse transcriptome v30 was downloaded and indexed.

* Kallisto 0.46.0 was used for mapping paired end reads to the transcriptome.

* Count data was loaded into R, followed by DE analysis with DESeq2.

* Enrichment analysis was performed with mitch, using REACTOME pathways.

* OS, R and package versions are shown in the session info section at the end of this report.


```{r,libs}

suppressPackageStartupMessages({
  library("reshape2")
  library("DESeq2")
  library("gplots")
  library("mitch")
  library("dplyr")
  library("biomaRt")
  library("kableExtra")
  library("beeswarm")
})

```

## Data

```{r,load}

tmp <- read.table("3col.tsv.gz",header=F)
x <- as.matrix(acast(tmp, V2~V1, value.var="V3", fun.aggregate = sum))
x <- as.data.frame(x)
txinfo <-  data.frame(do.call(rbind,strsplit(rownames(x),"\\|")))
colnames(txinfo) <- c("tx","GeneID","altGeneID","altTxID",
  "txname","genename","length","biotype")
x$GeneID <- paste(txinfo$GeneID , txinfo$genename)

# write out transcript table
x2 <- x
x2$GeneID = NULL
xrpm <- apply(x2, 2, function(x){x/sum(x,na.rm=T)}) * 1000000
colnames(xrpm) <- gsub("$","_RPM",colnames(xrpm))
colnames(x2) <- gsub("$","_counts",colnames(x2))
x2 <- cbind(xrpm,x2)
write.table(x2,file="Tx_RPM_count_table_biotype.tsv")

xx <- aggregate(. ~ GeneID, x , sum)
colnames(xx) <- gsub("mRNA_","",colnames(xx))
rownames(xx) <- xx$GeneID
xx$GeneID <- NULL
xx <- round(xx)
dim(xx)
head(xx)
write.table(x=xx,file="mRNA_count_matrix.tsv",sep="\t")

```

Now quantify the gene biotypes.
Starting with all samples.

```{r,biotype1,figure.width=12,figure.height=7}

txinfo2 <- txinfo
txinfo2$gene <- paste(txinfo2$GeneID , txinfo2$genename)
biotype <- unique(txinfo2[,c("gene","biotype")])
xxbiotype <- merge(xx,biotype,by.x=0,by.y="gene")
xxbiotype$Row.names=NULL

xxbiotype <- aggregate(. ~ biotype,xxbiotype,sum)
rownames(xxbiotype) <- xxbiotype$biotype
xxbiotype$biotype = NULL
# remove biotypes with very few reads
xxbiotype <- xxbiotype[which(rowMeans(xxbiotype)>10),]
# keep top 10 classes for chart
n=10
xxbiotype <- xxbiotype[tail(order(rowMeans(xxbiotype)),n),]
# make percent
xxbiotype2 <- apply(xxbiotype, 2, function(x){x*100/sum(x,na.rm=T)})
# create color palette:
library(RColorBrewer)
pal <- brewer.pal(n, "Paired")

par(mfrow=c(1,2))
par(mar=c(10,5,3,1))
barplot(as.matrix(xxbiotype2),col=pal,las=2)
plot.new()
legend("topleft", legend=rev(rownames(xxbiotype2)),fill=rev(pal),bg="white",cex=1)

#restore default setting
par(mar=c(5.1, 4.1, 4.1, 2.1))

```

Now with just the samples not treated with TGFb.

```{r,biotype2,figure.width=12,figure.height=7}

txinfo2 <- txinfo
txinfo2$gene <- paste(txinfo2$GeneID , txinfo2$genename)
biotype <- unique(txinfo2[,c("gene","biotype")])
xxbiotype <- merge(xx,biotype,by.x=0,by.y="gene")
xxbiotype$Row.names=NULL

xxbiotype <- aggregate(. ~ biotype,xxbiotype,sum)
rownames(xxbiotype) <- xxbiotype$biotype
xxbiotype$biotype = NULL
# remove TGFb samples
xxbiotype <- xxbiotype[,grep("TGF",colnames(xxbiotype),invert=TRUE)]

# remove biotypes with very few reads
xxbiotype <- xxbiotype[which(rowMeans(xxbiotype)>10),]
# keep top 10 classes for chart
n=10
xxbiotype <- xxbiotype[tail(order(rowMeans(xxbiotype)),n),]
# make percent
xxbiotype2 <- apply(xxbiotype, 2, function(x){x*100/sum(x,na.rm=T)})
# create color palette:
library(RColorBrewer)
pal <- brewer.pal(n, "Paired")

par(mfrow=c(1,2))
par(mar=c(10,5,3,1))
barplot(as.matrix(xxbiotype2),col=pal,las=2)
plot.new()
legend("topleft", legend=rev(rownames(xxbiotype2)),fill=rev(pal),bg="white",cex=1)

#restore default setting
par(mar=c(5.1, 4.1, 4.1, 2.1))

```


## Samplesheet

```{r,samplesheet}

#colnames(xx) <- sapply(strsplit(colnames(xx),"_"),"[[",1)
mysamples <- colnames(xx)
ko <- as.numeric(grepl("ko",mysamples))
tgf <- as.numeric(grepl("TGF",mysamples))
ss <- data.frame(mysamples,ko,tgf)
rownames(ss) <- mysamples
ss

```

## QC

TODO: rRNA carryover.

```{r,qc}

par(mar=c(5,8,3,1))
barplot(colSums(xx),horiz=TRUE,las=1,xlab="num reads",col=ss$cols)
grid()

```

## MDS plot

And correlation heat map.

```{r,mds}

par(mar=c(5,5,3,3))

mds <- cmdscale(dist(t(xx)))

plot(mds, xlab="Coordinate 1", ylab="Coordinate 2",
  type = "p",bty="n", cex=4 )

text(mds, labels=rownames(mds) ,col="black")

heatmap.2(cor(xx),trace="n",main="Pearson correlation heatmap",margin=c(8,8),cexRow=0.7,cexCol=0.7)

```

## Differential expression

Compare wt control to RAGE KO control.

```{r,func}

maplot <- function(de,contrast_name) {
  de <- de[which(!is.na(de$padj)),]
  sig <-subset(de, padj < 0.05 )
  up <-rownames(subset(de, padj < 0.05 & log2FoldChange > 0))
  dn <-rownames(subset(de, padj < 0.05 & log2FoldChange < 0))
  GENESUP <- length(up)
  GENESDN <- length(dn)
  DET=nrow(de)
  SUBHEADER = paste(GENESUP, "up, ", GENESDN, "down", DET, "detected")
  ns <-subset(de, padj > 0.05 )
  plot(log2(de$baseMean),de$log2FoldChange,
       xlab="log2 basemean", ylab="log2 foldchange",
       pch=19, cex=0.5, col="dark gray",
       main=contrast_name, cex.main=1)
  points(log2(sig$baseMean),sig$log2FoldChange,
         pch=19, cex=0.5, col="red")
  mtext(SUBHEADER,cex = 1)
}
make_volcano <- function(de,name) {
    de <- de[which(!is.na(de$padj)),]
    de$pvalue[which(de$pvalue==0)] <- 1e-320
    sig <- subset(de,padj<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,log2FoldChange>0))
    N_DN=nrow(subset(sig,log2FoldChange<0))
    DET=nrow(de)
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn", DET, "detected")
    plot(de$log2FoldChange,-log10(de$pval),cex=0.5,pch=19,col="darkgray",
        main=name, xlab="log2 FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$log2FoldChange,-log10(sig$pval),cex=0.5,pch=19,col="red")
}

```

```{r,de01}

ss1 <- subset(ss,tgf==0)
xx1 <- xx[,which(colnames(xx) %in% rownames(ss1) )]
dim(xx1)
xx1 <- xx1[which(rowMeans(xx1)>10),]
dim(xx1)

dds <- DESeqDataSetFromMatrix(countData = xx1 , colData = ss1 , design = ~ ko )
res <- DESeq(dds)
z <- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd))
dge <- as.data.frame(zz[order(zz$pvalue),])
dge[1:20,1:6] %>% kbl(caption = "Top gene expression differences") %>%
  kable_paper("hover", full_width = F)
d1up <- rownames(subset(dge,padj <= 0.05 & log2FoldChange > 0))
d1dn <- rownames(subset(dge,padj <= 0.05 & log2FoldChange < 0))

txinfo$GeneID <- sapply(strsplit(txinfo$GeneID,"\\."),"[[",1)

g <- sapply(strsplit(rownames(dge),"\\."),"[[",1)

dge$biotype <- sapply(g, function(x) {
  paste(unique(txinfo[txinfo$GeneID==x,"biotype"]),collapse=",") }
)

write.table(dge,file="mRNA_DE_ctrl_vs_rageko.tsv",quote=FALSE,sep="\t")

maplot(dge,"Cont1: Effect of RAGE KO")
make_volcano(dge,"Cont1: Effect of RAGE KO")

#agerdat <- assay(vsd)[grep("Ager",rownames(assay(vsd))),]
#par(mar=c(5,8,3,1))
#barplot(agerdat,horiz=TRUE,las=1,xlab="Normalised Ager expression",xlim=c(0,10))

```

Make a heatmap of the top 30 genes.


```{r,heatmapx9}

rpm1 <- apply(xx1, 2, function(x){x/sum(x,na.rm=T)}) * 1000000

agerdat <- rpm1[grep("Ager",rownames(rpm1)),]
par(mar=c(5,8,3,1))
barplot(agerdat,horiz=TRUE,las=1,xlab="Normalised Ager expression",xlim=c(0,16))

colfunc <- colorRampPalette(c("blue", "white", "red"))

rpm2 <- rpm1[which(rownames(rpm1) %in% rownames(head(dge,30))),]

heatmap.2(as.matrix(rpm2),margin=c(8, 22),cexRow=0.85,trace="none",
    cexCol=0.9,col=colfunc(20),scale="row")

rownames(rpm2) <- sapply(strsplit(rownames(rpm2)," "),"[[",2)

heatmap.2(as.matrix(rpm2),margin=c(8, 22),cexRow=0.85,trace="none",
    cexCol=0.9,col=colfunc(20),scale="row")


```

## Pathway analysis

First need ortholog table.

Downloaded separately from the ensembl biomart website because the R package code is timing out.

```{r,ortho2}

orth <- read.table("../../ref/mart_export.txt",sep="\t",header=TRUE)
orth <- orth[which(orth$Human.gene.name != ""),]
orth <- orth[,c("Gene.stable.ID","Human.gene.name")]
orth <- unique(orth)
rownames(dge) <- sapply(strsplit(rownames(dge),"\\."),"[[",1)

```

Run mitch.

See the other report called "mrna_analysis_mitch.html" for the detailed analysis.

```{r,mitch}

#download.file("https://reactome.org/download/current/ReactomePathways.gmt.zip", destfile="ReactomePathways.gmt.zip")
#unzip("ReactomePathways.gmt.zip")
# downloaded 19th August 2022
genesets <- gmt_import("ReactomePathways.gmt")

y <- mitch_import(dge, DEtype="deseq2",geneTable=orth)

head(y)

res <- mitch_calc(y, genesets, priority="effect",cores=16)


if (! file.exists("mrna_analysis_mitch.html")) {
  mitch_report(res, "mrna_analysis_mitch.html")
}

head(res$enrichment_result,20) %>%
  kbl(row.names=FALSE,caption = "Top pathways") %>%
  kable_paper("hover", full_width = F)

write.table(res$enrichment_result,file="mitch_pathways.tsv",sep="\t")

```

## MiR-214

We have a list of mir-214 targets.

```{r,blah}

mir214_targ <- readLines("miR214_targets.txt")

mir214_targ <- list(mir214_targ)

names(mir214_targ) <- "mir214_targ"

res2 <- mitch_calc(y, mir214_targ, priority="effect",cores=16)

res2$enrichment_result

if (! file.exists("mrna_analysis_mitch_mir214.html")) {
  mitch_report(res2, "mrna_analysis_mitch_mir214.html")
}

par(mfrow=c(2,1))
MIN=min(res2$ranked_profile[,1])
MAX=max(res2$ranked_profile[,1])
beeswarm(res2$detailed_sets$mir214_targ,pch=19,
  lty="blank",cex=0.45,horiz=TRUE,ylim=c(MIN,MAX),
  main="miR-214 targets",
  xlab="position in rank (Ctrl vs RAGE KO)"
)
abline(v=0,lwd=2,lty=2,col="red")
abline(v=MIN,lwd=2,lty=2,col="black")
abline(v=MAX,lwd=2,lty=2,col="black")
grid()

par(mfrow=c(1,1))

```

It appears that they are upregulated in this contrast.

## Reactome analysis of miR-214 targets

Here we are going to try running a 1 sample t-test.

1. Extract the miR214 targets from the mitch rank.

2. For each Reactome, get the set that intersects with miR-214 targets.

3. Run 1-sample t-test to determine enrichment.

4. Collect and format results.

5. Present chart and enriched genes for the top 10 pathways

```{r,1sttest}

head(res2$ranked_profile)
MIN=min(as.vector(res2$ranked_profile))
MAX=max(as.vector(res2$ranked_profile))

str(res2$detailed_sets)
head(res2$detailed_sets[[1]])
tail(res2$detailed_sets[[1]])

hist(res2$detailed_sets[[1]],xlab="gene rank")
hist(res2$detailed_sets[[1]],xlab="gene rank",breaks=100)
y <- res2$detailed_sets[[1]]

names(genesets[lapply(genesets,length) >500])

# 1 sample t-test self contained test
# z = profile
# i = geneset index
# gsl = gene set library
# n = min number fo genes overlap=5
t1 <- function(y,i,gsl,n=5) {
  gs <- gsl[[i]]
  gsname <- names(gsl)[i]
  z <- y[names(y) %in% gs]
  len=length(z)
  if (len >= n ) {
    tres=t.test(z)
    pval=tres$p.value
    zmean=mean(z)
    zmedian=median(z)
    zcil=tres$conf.int[1]
    zciu=tres$conf.int[2]

    res = list(gsname=gsname,"p-value"=pval,"mean"=zmean,"median"=zmedian,
      "lower conf interval"=zcil,"upper conf interval"=zciu,numgenes=len)

    return(res)
  }
}

# test it
t1(y,143,genesets)

# run it for all
res <- lapply(1:length(genesets), function(i) {
  unlist(t1(y,i,gsl,n=5))
})

# format the results
out <- t(do.call(cbind,res))
rownames(out) <- out[,1]
out <- out[,2:ncol(out)]
out <- as.matrix(out)
out2 <- apply(out,2,as.numeric)
rownames(out2) <- rownames(out)
out <- as.data.frame(out2)
out$padj <- p.adjust(out$`p-value`)
out <- out[order(out$`p-value`),]

```

## Now we can take a look at the results

Top significant results.

Take this table with a grain of salt because small p-values don't always mean a biologically
meaningful observation.
For example, some of these are very large sets of genes with only a slight change.

```{r,t1top}

head(out,20) %>% kbl(caption = "Top significant pathways") %>%
  kable_paper("hover", full_width = F)

```

Biggest effect size (upregulated) with FDR<0.05, ranked by median.

```{r,t1up}

sigup <- subset(out,padj<0.05 & median>0)
nrow(sigup)

upreg <- head(sig[order(-sigup$median),],20)

upreg %>% kbl(caption = "Top upregulated pathways") %>%
  kable_paper("hover", full_width = F)

```

Biggest effect size (downregulated) with FDR<0.05, ranked by median.

```{r,t1dn}

sigdn <- subset(out,padj<0.05 & median<0)
nrow(sigdn)

dnreg <- head(sig[order(sigdn$median),],20)

dnreg %>% kbl(caption = "Top downregulated pathways") %>%
  kable_paper("hover", full_width = F)

```

There were no downregulated sets that met the FDR threshold.

Take a look at a few top sets.

First with some histograms.

```{r,t1detail,fig.height=5}

gsnames <- head(rownames(upreg),10)
for (gsname in gsnames ) {
  gs <- genesets[[gsname]]
  z <- y[names(y) %in% gs]
  hist(z,xlim=c(MIN,MAX),breaks=10,xlab="generank",main=gsname)
}

for (gsname in gsnames ) {
  gs <- genesets[[gsname]]
  print(gsname)
  z <- y[names(y) %in% gs]
  z <- z[order(-z)]
  print(z)
}

```

## SessionInfo

```{r,session}

sessionInfo()

```
